azure-iot-device/azure/iot/device/provisioning/abstract_provisioning_device_client.py

```diff
@@ -32,30 +32,75 @@ def __init__(self, provisioning_pipeline):
-                security_client, X509SecurityClient
-            ):
-                mqtt_provisioning_pipeline = ProvisioningPipeline(security_client)
-                return cls(mqtt_provisioning_pipeline)
-                # TODO : other instances of security provider can also be checked before creating mqtt and client
-            else:
-                raise ValueError(
-                    "A symmetric key security client or a X509 security client must be provided for MQTT"
-                )
+            security_client = SymmetricKeySecurityClient(
+                provisioning_host, registration_id, id_scope, symmetric_key
+            )
+            mqtt_provisioning_pipeline = ProvisioningPipeline(security_client)
+            return cls(mqtt_provisioning_pipeline)
+        else:
+            raise NotImplementedError(
+                "A symmetric key can only create symmetric key security client which is compatible "
+                "only with MQTT protocol.Any other protocol has not been implemented."
+            )
 
+    @classmethod
+    def create_from_x509_certificate(
+        cls, provisioning_host, registration_id, id_scope, x509, protocol_choice=None
+    ):
+        """
+        Create a client which can be used to run the registration of a device with provisioning service
+        using X509 certificate authentication.
+        :param provisioning_host: Host running the Device Provisioning Service. Can be found in the Azure portal in the
+        Overview tab as the string Global device endpoint
+        :param registration_id: The registration ID is used to uniquely identify a device in the Device Provisioning Service.
+        The registration ID is alphanumeric, lowercase string and may contain hyphens.
+        :param id_scope: The ID scope is used to uniquely identify the specific provisioning service the device will
+        register through. The ID scope is assigned to a Device Provisioning Service when it is created by the user and
+        is generated by the service and is immutable, guaranteeing uniqueness.
+        :param x509: The x509 certificate, To use the certificate the enrollment object needs to contain cert (either the root certificate or one of the intermediate CA certificates).
+        If the cert comes from a CER file, it needs to be base64 encoded.
+        :param protocol_choice: The choice for the protocol to be used. This is optional and will default to protocol MQTT currently.
+        :return: A ProvisioningDeviceClient which can register via Symmetric Key.
+        """
+        if protocol_choice is not None:
+            protocol_name = protocol_choice.lower()
+        else:
+            protocol_name = "mqtt"
+        if protocol_name == "mqtt":
+            security_client = X509SecurityClient(provisioning_host, registration_id, id_scope, x509)
+            mqtt_provisioning_pipeline = ProvisioningPipeline(security_client)
+            return cls(mqtt_provisioning_pipeline)
```

Comment:
>return cls(mqtt_provisioning_pipeline) [](start = 12, length = 38)

This feels wrong.  Why don't you allocate the ProvisiongPipeline object inside the provisioning client initializer?  It feels like the initializer should accept a security client and then allocate the pipeline.

Reply:
never mind, I found a comment explaining this in carter's PR.  

---
In reply to: [294447595](https://github.com/Azure/azure-iot-sdk-python-preview/pull/109#discussion_r294447595) [](ancestors = 294447595)